realm-transformer/src/main/kotlin/io/realm/transformer/RealmOptionalAPITransformer.kt

```diff
@@ -0,0 +1,88 @@
+import io.realm.annotations.internal.OptionalAPI
+import org.slf4j.LoggerFactory
+import java.util.HashSet
+
+class RealmOptionalAPITransformer : Transform() {
+
+    private val logger = LoggerFactory.getLogger("realm-logger")
+
+    override fun getName(): String? = "RealmOptionalAPITransformer"
+
+    override fun isIncremental(): Boolean = false
+
+    override fun getInputTypes(): MutableSet<ContentType>? = ImmutableSet.of(DefaultContentType.CLASSES)
+
+    override fun getScopes(): MutableSet<Scope>? = ImmutableSet.of(Scope.EXTERNAL_LIBRARIES)
+
+    override fun getReferencedScopes(): MutableSet<Scope>? = ImmutableSet.of(Scope.PROJECT, Scope.PROJECT_LOCAL_DEPS,
+            Scope.SUB_PROJECTS, Scope.SUB_PROJECTS_LOCAL_DEPS, Scope.EXTERNAL_LIBRARIES)
+
+    override fun transform(context: Context?,
+                           inputs: MutableCollection<TransformInput>?,
+                           referencedInputs: MutableCollection<TransformInput>?,
+                           outputProvider: TransformOutputProvider?,
+                           isIncremental: Boolean) {
+
+        val classNames = HashSet<String>()
+        inputs!!.appendAllToClassNames(classNames)
+        val refClassNames = HashSet<String>()
+        referencedInputs!!.appendAllToClassNames(refClassNames)
+        val classPool = createClassPool()
+        inputs.appendAllToClassPool(classPool)
+
+        classNames.filter { it.startsWith("io.realm.") }.forEach {
+            classPool.get(it).declaredMethods.forEach {
+                val optionalAPIAnnotation = it.getAnnotation(OptionalAPI::class.java) as? OptionalAPI
+                val dependenciesList = optionalAPIAnnotation?.dependencies?.toList()
+
+                if (optionalAPIAnnotation == null) {
+                    logger.info("${it.declaringClass.name} ${it.name} doesn't have @OptionalAPI annotation.")
+                } else if (dependenciesList == null || dependenciesList.size == 0) {
+                    throw IllegalArgumentException("${it.name} doesn't have proper dependencies: " +
+                            "${optionalAPIAnnotation.dependencies}.")
+                } else if (!refClassNames.containsAll(dependenciesList)) {
+                    // Doesn't have enough dependencies, remove the API
+                    logger.info("${it.declaringClass.name} ${it.name} will be removed since some of the dependencies " +
+                            "in $dependenciesList don't exist.")
+                    it.declaringClass.removeMethod(it)
+                } else {
+                    logger.info("${it.declaringClass.name} ${it.name} has all dependencies in $dependenciesList.")
+                }
+            }
+        }
+
+        // Create outputs
+        classNames.forEach {
+            val ctClass = classPool.getCtClass(it)
+            ctClass.writeFile(outputProvider!!.getContentLocation(
+                    "realm-optional-api", inputTypes, scopes, Format.DIRECTORY).canonicalPath)
```

Comment:
Isn't "realm-optional-api" a constant? So perhaps define it at the top of the file. Kotlin doesn't really support static fields but perhaps just `val NAME = "realm-optional-api"` ? Not entirely sure what would be idiomatic there.